### PR TITLE
print stack trace for FATAL log.

### PR DESCRIPTION
**Pull request by csukuangfj**

### CODE REVIEW

**Comment by csukuangfj** [comment]
ready for review.

**Comment by danpovey** [comment]
LGTM, thanks!

File: k2/csrc/log.h [hunk 1]
```diff
@@ -93,6 +96,11 @@ class Logger {
       __assert_fail("Some bad things happened", filename_, line_num_,
                     func_name_);
 #else
+      std::string stack_trace = GetStackTrace();
```
Comment (qindazhu) [inline k2/csrc/log.h:99]: Can you guys build successfully with the latest master branch? it seems to me `GetStackTrace()` is defined in `log.cu`, but `log.h` is referenced in those /host code, so I will get error 
```
bin/ld: CMakeFiles/weights_test.dir/weights_test.cc.o: in function `k2::internal::Logger::~Logger()':
/ceph-hw/k2/k2/csrc/log.h:99: undefined reference to `k2::internal::GetStackTrace()'

```
I suppose @csukuangfj has definitely built successfully before committing code?

**Comment by csukuangfj** [comment]
Yes, you can check that GitHub actions build successfully.



Sent from myMail for iOS


Wednesday, 18 November 2020, 19:58 +0800 from notifications@github.com  <notifications@github.com>:

**Comment by csukuangfj** [comment]
Will create a fix for it



Sent from myMail for iOS


Wednesday, 18 November 2020, 19:58 +0800 from notifications@github.com  <notifications@github.com>:

**Comment by qindazhu** [comment]
I wonder how does this happen? As when building /host code, I was thinking it can't find the definition of GetStackTrace in log.cu, it's strange to me it can build successfully

**Comment by csukuangfj** [comment]
@qindazhu 

Can you create a separate cmake target for `log`? E.g.,

```cmake

add_library(k2log  log.cu)
if(K2_HAVE_EXECINFO_H)
if(K2_HAVE_EXECINFO_H)
  target_compile_definitions(k2log PRIVATE K2_HAVE_EXECINFO_H=1)
endif()

if(K2_HAVE_CXXABI_H)
  target_compile_definitions(k2log PRIVATE K2_HAVE_CXXABI_H=1)
endif()
``` 

And let `k2fsa` and `k2context` depend on `k2log`?

GitHub actions build only `_k2` and `k2context`; they do not build `k2fsa`.

I will update `GitHub actions`.

**Comment by csukuangfj** [comment]
I did not use `make -j` while creating this pull request. Only `make _k2` was tested locally.

**Comment by qindazhu** [comment]
> GitHub actions build only _k2 and k2context; they do not build k2fsa.
Ha, OK!

**Comment by qindazhu** [comment]
Maybe it's better if you can make a PR to fix it,  I tried to add depencies with add_library but it still failed.

**Comment by csukuangfj** [comment]
Ok. Will fix it one hour later.



Sent from myMail for iOS


Wednesday, 18 November 2020, 20:49 +0800 from notifications@github.com  <notifications@github.com>:

**Comment by csukuangfj** [comment]
> I tried to add dependencies with target_link_libraries but it still failed.

By default, `log.cu` uses a different ABI with `libk2fsa.so`, so even if the function name is `GetStackTrace`,
the name after mangling is different, resulting in the link error you encountered above.

Fixed in #362

**Pull request merged**